### ISSUE TITLE
Vertices with NaNs when meshing box

### ISSUE THREAD

**Issue by julianschuler** [mkeeter/fidget#15]
Hi,

I tried implementing a box primitive (adapted from [Inigo Quilez article](https://iquilezles.org/articles/distfunctions)) and noticed that meshing it results in a lot of vertices with NaNs.

I tested this with both the latest main and v0.2.0, with and without the JIT feature as well as different box sizes and octree evaluation depths. Here is a minimum working example reproducing this issue:

```rust
use fidget::{
    context::{IntoNode, Node},
    eval::MathShape,
    jit::JitShape,
    mesh::{Octree, Settings},
    Context, Error,
};

pub struct BoxShape {
    size: f64,
}

impl BoxShape {
    pub fn new(size: f64) -> Self {
        Self { size }
    }
}

impl IntoNode for BoxShape {
    fn into_node(self, context: &mut Context) -> Result<Node, Error> {
        let x = context.x();
        let y = context.y();
        let z = context.z();

        let abs_x = context.abs(x)?;
        let abs_y = context.abs(y)?;
        let abs_z = context.abs(z)?;

        let qx = context.sub(abs_x, self.size)?;
        let qy = context.sub(abs_y, self.size)?;
        let qz = context.sub(abs_z, self.size)?;

        let max_x = context.max(qx, 0.0)?;
        let max_y = context.max(qy, 0.0)?;
        let max_z = context.max(qz, 0.0)?;

        let x2 = context.square(max_x)?;
        let y2 = context.square(max_y)?;
        let z2 = context.square(max_z)?;

        let sum = context.add(x2, y2)?;
        let sum = context.add(sum, z2)?;

        let outer = context.sqrt(sum)?;

        let max_elem = context.max(qx, qy)?;
        let max_elem = context.max(max_elem, qz)?;

        let inner = context.min(max_elem, 0.0)?;

        context.add(outer, inner)
    }
}

fn main() -> Result<(), Error> {
    let mut context = Context::new();

    let box_shape = BoxShape::new(0.5);
    let root = box_shape.into_node(&mut context)?;

    let settings = Settings {
        threads: 12,
        min_depth: 4,
        max_depth: 4,
    };

    let shape = JitShape::new(&context, root)?;
    let mesh = Octree::build(&shape, settings).walk_dual(settings);

    // Count number of the vertices with non-finite values
    let invalid_vertices = mesh
        .vertices
        .iter()
        .filter(|vertex| vertex.iter().any(|p| !p.is_finite()))
        .count();

    dbg!(invalid_vertices, mesh.vertices.len());

    Ok(())
}
```

The output of this example is the following (expected `invalid_vertices` to be 0):

```
[src/main.rs:77] invalid_vertices = 296
[src/main.rs:77] mesh.vertices.len() = 590
```

The error seems to be related to the QEF solving (in earlier versions of fidget I noticed vertices with components that were not NaN, but had implausible high values >100000 for their components).

I am not that knowledgeable in the QEF math though, please let me know if I can help in any other way resolving this issue!

**Comment by mkeeter** [comment]
For reference, it looks like this is the "round box" example, which is
```glsl
float sdRoundBox( vec3 p, vec3 b, float r )
{
  vec3 q = abs(p) - b;
  return length(max(q,0.0)) + min(max(q.x,max(q.y,q.z)),0.0) - r;
}
```

The problem is that this expression doesn't ensure that its partial derivatives are well-behaved.  That makes sense for raycasting (where you only care about the distance), but Fidget uses the partial derivatives for vertex positioning.

Specifically, `length(max(q, 0.0))` requires taking a square root, and the partial derivatives of the square root at [blows up at 0.0](https://www.wolframalpha.com/input?i=d%2Fdx+sqrt%28f%28x%2C+y%2C+z%29%29).  Having NaN values for gradients corrupts the QEF matrices, which leads to nonsense vertex positions.

One dumb fix is to clamp to 0.0001 instead of 0.0, but that's kinda gross.  A cleaner fix would be to find an equation for a rounded box which has well-behaved gradients.

**Comment by mkeeter** [comment]
Taking a quick look, it seems like `libfive` is less rigorous in its handling of `d/dx sqrt(x)` at `x = 0`, special-casing this condition to zero:
```cpp
        case Opcode::OP_SQRT:
            for (Eigen::Index i=0; i < od.rows(); ++i)
                od.row(i) = (av < 0 || ad.row(i) == 0).select(
                    Eigen::Array<float, 1, Eigen::Dynamic>::Zero(1, count_simd),
                    ad.row(i) / (2 * ov));
            break;
```

Right now, I'm not planning to change Fidget's behavior to match libfive, but I may reconsider at some point!

(I also did a quick experiment with skipping NaN values at the QEF level, but that didn't seem to help this model)
